lighthouse-core/gather/computed/main-resource.js

```diff
@@ -0,0 +1,27 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const ComputedArtifact = require('./computed-artifact');
+const HTTP_REDIRECT_CODE_LOW = 300;
+const HTTP_REDIRECT_CODE_HIGH = 399;
+
+class MainResource extends ComputedArtifact {
+  get name() {
+    return 'MainResource';
+  }
+
+  /**
+   * @param {!Array<!WebInspector.NetworkRequest>} networkRecords
+   * @return {?WebInspector.NetworkRequest}
+   */
+  compute_(networkRecords) {
+    return networkRecords.find(record => record.statusCode < HTTP_REDIRECT_CODE_LOW ||
```

Comment:
I may be missing something obvious, but I'm not sure how this check is grabbing the main resource? :)

Reply:
oh, is it that `networkRecords` is sorted? I guess it would have to be the first non-redirect. Would it be better to have a stronger check for this?

Reply:
Hm, just found @wardpeet PR that deals with redirects too -
 https://github.com/GoogleChrome/lighthouse/pull/3308  . We could also use `CriticalRequestChains` here and check for first resource that doesn't have `redirected` key (instead of using status codes).

Reply:
we may want to avoid using `CriticalRequestChains` as they're due for a refactor after more fast-mode dependency graph stuff lands.

@paulirish may want to weigh in on what he imagines as a useful artifact here, but going way back to #605 he had the idea or a redirect chain artifact, which would work for both here and #3308 (and be much simpler to iterate over compared to the critical request chains data structure).

Reply:
I'm also like 80% sure @paulirish has told me the definitive way to determine the main request record, but I don't recall what that was. I defer to your research in the meantime, but relying on array order alone does make me a little nervous.

Lighthouse inserts into the networkRecords array based on when a request is finished, and it seems like it's possible (maybe? I have no evidence either way :) that the preload scanner (or something else) could trigger a resource download that finishes before the main page finishes loading, and so would be inserted before it and trick the statusCode check. The other checks would help work around this.

Walking the redirect chain (and maybe explicitly matching redirected URLs with `redirectSource` like [driver does for monitoring redirects](https://github.com/GoogleChrome/lighthouse/blob/3f247a947efd76e5d8c3dbac2e590614480ea541/lighthouse-core/gather/driver.js#L569-L581)) also seems like it would solve this problem well

Reply:
> Lighthouse inserts into the networkRecords array based on when a request is finished

whoops, [I'm wrong](
https://github.com/GoogleChrome/lighthouse/blob/3f247a947efd76e5d8c3dbac2e590614480ea541/lighthouse-core/lib/network-recorder.js#L55). As of #2197 they are inserted as the request begins, so this may be fine

Reply:
> so this may be fine.
so… we are back to first non-redirect heuristic?

> redirect-chain artifact

NetworkRequest's have `redirects` property that exposes chain of redirects that led to that resource. So, you can do `mainResoruce.redirects` and get this:
<img width="1351" alt="screen shot 2017-09-25 at 00 00 32" src="https://user-images.githubusercontent.com/985504/30787309-da0e2ac8-a185-11e7-9bbe-b8a51d65be3d.png">
this should be enough for #3308 (@wardpeet WDYT?). We can rework it to `redirect-chain` if needed, but IMO `mainResource.redirects` works well.

Reply:
> I'm also like 80% sure @paulirish has told me the definitive way to determine the main request record, but I don't recall what that was

looking in DT source i see some really basic approaches for finding "main resources"

```js
      if (networkManager && request.url() === networkManager.target().inspectedURL() &&
          request.resourceType() === Common.resourceTypes.Document)
```

```js
(roughly what resourcetreemodel is doing...)
let url;
// _resourcesMap populated ...
target.on(SDK.ResourceTreeModel.Events.FrameNavigated, framePayload => url = framePayload.url);
var mainResource = this._resourcesMap[url];
```

basically just looking at URL match.. and either excluding redirects via selecting just Resources or by only considering type==document.  ok then.


so tldr there is no definitive way. what you have here lgtm